### PR TITLE
CNS-788: Subscription module README

### PR DESCRIPTION
**Pull request by shleikes**

### CODE REVIEW

**Comment by shleikes** [comment]
CNS-783

File: x/subscription/README.md [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+   - Else, if auto-renewal is enabled:
+     - Attempt to renew the subscription for another month.
+     - If renewal fails, remove the expired subscription.
+   - If there is no future subscription or auto-renewal, remove the expired subscription.
+
+### Subscription Upgrade
+
+A subscription can be upgraded to a more expensive plan.
+
+That also means, that if a consumer buys a plan, and then decides to downgrade their plan, the way to do it is either buy more months in the cheaper subscription, so the amount that the consumer will pay will be larger than the one using the same command as shown above, with the only caveat - the `plan-index` must be different than the currently active subscription's plan, and, it must be with a higher price.
```
Comment (oren-lava) [inline x/subscription/README.md:121]: I really didn't understand this paragraph. can you rephrase it? also, i thought downgrading is not possible

File: x/subscription/README.md [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+
+Tokens are deducted from the creator's account immediately, and the new plan becomes effective in the next epoch.  
+The old subscription will still be saved in the fixation store, for buffering purposes (the amount of blocks that the subscription will be saved for, is determined in the `BlocksToSave` function, under the [EpochStorage](https://github.com/lavanet/lava/blob/main/x/epochstorage/README.md) module).
```
Comment (oren-lava) [inline x/subscription/README.md:124]: not sure that this is relevant to the reader (it's TMI)

File: x/subscription/README.md [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+
+More ways to upgrade are by making an [Advance Purchase](#advance-purchase) or enabling [Auto Renewal](#auto-renewal).
+
+### Subscription Renewal
```
Comment (oren-lava) [inline x/subscription/README.md:128]: why separate subscription renewal and auto renewal? put them under the same section

File: x/subscription/README.md [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+lavad tx subscription buy --advance-purchase [plan-index] [optional: consumer] [optional: duration(months)] [flags]
+```
+
+This will create the `FutureSubscription` object inside the `Subscription` object, as can be seen above.
+The new subscription will be triggered once the current active subscription expires.
+
+Users can decide on the duration of this future subscription, and they'll pay the full amount for the entire period immediately.
+If the plan changes during that period, users who bought this plan before it changed using the `--advance-purchase` flag, they won't be affected by that change.
+
+If a user tries to replace the future subscription with another, the new plan's price must be higher, considering the amount of days bought.
```
Comment (oren-lava) [inline x/subscription/README.md:147]: bough -> bought

File: x/subscription/README.md [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+Meaning, if user originally bough X days of a plan with price A, and now wants to advance purchase Y days of a different plan with price B, than the following must be suffice:
+
+$$
+Y * B > X * A
+$$
+
+### Auto Renewal
+
+Users can decide to if they want the subscription to auto-renew itself every month, using the subscription buy transaction command:
```
Comment (oren-lava) [inline x/subscription/README.md:156]: fix to: Users can make the subscription auto-renew itself..

Also, using the auto-renewal transaction (not buy TX)

File: x/subscription/README.md [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+
+```bash
+lavad tx subscription auto-renewal [true/false] [optional: plan-index] [optional: consumer] [flags]
+```
+
+Here, users can decide to renew their subscription to any plan index, wether it's cheaper or more expensive than the current active subscription's plan.
```
Comment (oren-lava) [inline x/subscription/README.md:162]: wether -> whether

File: x/subscription/README.md [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+
+The tokens will be taken from the user's account only when at the end of each month, prior to renewing the subscription, and for one month only.
```
Comment (oren-lava) [inline x/subscription/README.md:164]: only when at the.. -> only at the

File: x/subscription/README.md [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+| `next-to-month-expiry` | none                  | Shows the subscriptions with the closest month expiry          |
+| `params`               | none                  | Shows the parameters of the module                             |
+
+## Transactions
+
+All the transactions below require setting the `--from` flag and gas related flags.
+
+The subscription module supports the following transactions:
+
+| Transaction    | Arguments                                                                               | What it does                                  |
```
Comment (oren-lava) [inline x/subscription/README.md:191]: in the "what it does" section, specify when the TX will be applied (immediately? next epoch?)

**Review by oren-lava — CHANGES_REQUESTED** [review]

File: x/subscription/README.md [hunk 9]
```diff
@@ -0,0 +1,202 @@
+# `x/subscription`
+
+## Abstract
+
+This document specifies the subscription module of Lava Protocol.
+
+The subscription module is responsible for managing Lava's consumer subscription.  
+To use Lava, consumers must purchase a subscription.  
+The subscription operates on a monthly basis, starting from the next block after purchase and ending one month later.  
+If a user purchases a subscription for more than one month, it will reset at the end of each month until its expiration.
```
[the remark this answers is not in the document]
Comment (shleikes) [reply x/subscription/README.md:10]: > why write reset here and explain it a sentence later?

What do you mean by that?

> I think you should note that the subscription has different limitations enforced by the effective policy. One of the limitations is the allocated monthly CU. When a month passes, the monthly CU counter is reset (or something like that)

How about this:
```
Subscriptions are subject to limitations set by the current policy, including a cap on monthly Compute Units (CUs).  
These CUs reset with each new subscription month, ensuring users get a fresh allocation based on their plan.
```

File: x/subscription/README.md [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+### Subscription
+
+In order for a consumer to purchase a subscription, they first must choose a plan to use.  
+After choosing a plan, a consumer can perform the purchase using the transaction command:
+
+```bash
+lavad tx subscription buy [plan-index] [optional: consumer] [optional: duration(months)] [flags]
+```
+
+More on the `buy` transaction is under the [Transactions](#transactions) part under Buy.
```
[the remark this answers is not in the document]
Comment (shleikes) [reply x/subscription/README.md:48]: > for more details about the buy transaction, see below

I feel like this is basically what I wrote, isn't it?

File: x/subscription/README.md [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+- [Queries](#queries)
+- [Transactions](#transactions)
+- [Proposals](#proposals)
+
+## Concepts
+
+### Subscription
+
+In order for a consumer to purchase a subscription, they first must choose a plan to use.  
+After choosing a plan, a consumer can perform the purchase using the transaction command:
```
[the remark this answers is not in the document]
Comment (shleikes) [reply x/subscription/README.md]: Fixed in 5d277900b55147245d0c6da83b214ef6e68246af

File: x/subscription/README.md [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+The subscription operates on a monthly basis, starting from the next block after purchase and ending one month later.  
+If a user purchases a subscription for more than one month, it will reset at the end of each month until its expiration.  
+When a subscription is reset, the monthly left CUs is set to the plan's monthly CU.
+
+A subscription can be renewed, manually or automatically.  
+Additionally, a subscription plan can be bought in advance for months ahead.
+
+There are 2 other concepts that the subscription is connected to:
+
+- [Plans](https://github.com/lavanet/lava/blob/main/x/pairing/README.md)
```
[the remark this answers is not in the document]
Comment (shleikes) [reply x/subscription/README.md]: Fixed in 041db8260c9355cad4ad2cad9309a5f9e96d1d1a

File: x/subscription/README.md [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+
+### Advance Month
+
+One of the primary processes within this module involves the function `advanceMonth`, which is invoked upon the expiration of a timer set by the subscription. This function holds significance as it manages the logic that occurs at the conclusion of each month of a subscription.
+
+Here's a high-level flow of the function:
+
+1. **Update the CU Tracker Timer:**
+
+   - Add a CU tracker timer for the subscription.
```
[the remark this answers is not in the document]
Comment (shleikes) [reply x/subscription/README.md]: Fixed in a7aac97a5d47ae53d57d6f06e37c0afe4b2e17bf

File: x/subscription/README.md [hunk 14]
```diff
@@ -0,0 +1,202 @@
...
+
+2. **Check Subscription Duration:**
+
+   - If the subscription’s remaining duration (`DurationLeft`) is zero:
```
[the remark this answers is not in the document]
Comment (shleikes) [reply x/subscription/README.md:99]: So, should I remove the pseudo code completely or just change this line?

File: x/subscription/README.md (hunk 1 above)
Comment (shleikes) [reply x/subscription/README.md:121]: Downgrading is not possible, but there are ways to achieve it in an indirect way, which I explain in this section. 

How's this:
```
While direct downgrading of a subscription plan is not officially supported, consumers seeking a lower-tier plan can still achieve a similar effect. This can be done by purchasing additional months under a less expensive plan. It's important to note that the new plan's plan-index must differ from the currently active one, and the total payment for the extended period under the cheaper plan should exceed the amount of the original, higher-priced plan. This workaround allows consumers to effectively 'downgrade' their subscription within the system's constraints
```

File: x/subscription/README.md (hunk 2 above)
Comment (shleikes) [reply x/subscription/README.md]: Removed in df4e95e548a27ddbc43d5823e858c949e7b9d35d

File: x/subscription/README.md (hunk 3 above)
Comment (shleikes) [reply x/subscription/README.md:126]: Combined in 7ef2ecf36129e8175e2ff9e0dd04eef924e6d1f2

File: x/subscription/README.md (hunk 5 above)
Comment (shleikes) [reply x/subscription/README.md]: Fixed in 7ef2ecf36129e8175e2ff9e0dd04eef924e6d1f2

File: x/subscription/README.md (hunk 4 above)
Comment (shleikes) [reply x/subscription/README.md:156]: I think this is already bought, isn't it?

File: x/subscription/README.md (hunk 6 above)
Comment (shleikes) [reply x/subscription/README.md]: Fixed in https://github.com/lavanet/lava/commit/7ef2ecf36129e8175e2ff9e0dd04eef924e6d1f2

File: x/subscription/README.md (hunk 7 above)
Comment (shleikes) [reply x/subscription/README.md]: Fixed in https://github.com/lavanet/lava/commit/7ef2ecf36129e8175e2ff9e0dd04eef924e6d1f2

File: x/subscription/README.md (hunk 8 above)
Comment (shleikes) [reply x/subscription/README.md]: Added in c26d2c9b0f58a4e7d315762a912e39e616bee8e1

File: x/subscription/README.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/subscription/README.md:10]: the change you suggested look good

I meant that you write `If a user purchases a subscription for more than one month, it will reset at the end of each month until its expiration.` and then in another sentence: `When a subscription is reset, the monthly left CUs is set to the plan's monthly CU.`

So i didnt understand why didn't you just write something like: "in the end of the month, the monthly left CU is reset" which is simpler

File: x/subscription/README.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/subscription/README.md:47]: It just felt a little too explicit. like: for the buy command go to the command table and look at the buy row. Obviously if someone needs more info he'll look at the buy row. You just need to refer to the correct table.

But it's not crucial, you can keep it as is if you'd like

File: x/subscription/README.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/subscription/README.md:98]: I think you should remove it completely and explain in more general terms

File: x/subscription/README.md (hunk 1 above)
Comment (oren-lava) [reply x/subscription/README.md:120]: yeah, this is much clearer.
Just a question, this is kind of a "weird trick" to achieve something pretty common. Is this behavior wanted? why would we limit the auto-renewal mechanism like this? if downgrade is possible and not contradicting other parts of the code, why won't we let consumers buy cheaper plans with any duration?

Comment (oren-lava) [reply x/subscription/README.md:120]: actually, looking at the code downgrading is not possible in any way. When you upgrade the plan, you check the plan price alone (without multiplying by the duration), so downgrading won't work

File: x/subscription/README.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (shleikes) [reply x/subscription/README.md:10]: Got it, how's this:
>If a user purchases a subscription for more than one month, it will reset at the end of each month until its expiration, with the monthly remaining Compute Units (CUs) being reset to the plan's monthly CU allocation.

[the remark this answers is not in the document]
Comment (oren-lava) [reply x/subscription/README.md:10]: yeah, that's good

File: x/subscription/README.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (shleikes) [reply x/subscription/README.md]: Agreed. Removed in 011f46b2c3eb33a95483503900b308647cdbe068

File: x/subscription/README.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (shleikes) [reply x/subscription/README.md]: Fixed in 2d67e86bff8057cd2d2c51bdbf80951cea403795

File: x/subscription/README.md (hunk 1 above)
Comment (shleikes) [reply x/subscription/README.md]: Exactly. Currently it is not supported, simply because it's too complicated to implement, alongside the face that it is not really a must-have feature.

Rephrase the paragraph in 0209b1a3defe1341632864c7780f1adb17052c70

File: x/subscription/README.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (shleikes) [reply x/subscription/README.md:10]: Awesome, fixed in 151e51f0462996dd83f355d61c69c3b84947dab5

[the remark this answers is not in the document]
Comment (shleikes) [reply x/subscription/README.md]: Awesome, fixed in 151e51f0462996dd83f355d61c69c3b84947dab5

File: x/subscription/README.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/subscription/README.md]: I would keep the referral to the pairing module for the CU tracker (because it's a term that you didn't mention before). Other than that, looks great

[the remark this answers is not in the document]
Comment (shleikes) [reply x/subscription/README.md]: Added a reference in 9bee5553d7176c3b1fb01c4ec76df0af09c6b503

**Review by oren-lava — APPROVED** [review]

File: x/subscription/README.md [hunk 15]
```diff
@@ -0,0 +1,180 @@
...
+
+There could only be a single active subscription for a consumer.  
+When a consumer purchase a subscription, the tokens are transferred from their account immediately.  
+Those tokens are later on distributed among the providers that served that consumer, and the delegators that staked those providers.
+
+### Advance Month
+
+One of the primary processes within this module involves the function `advanceMonth`, which is invoked upon the expiration of a timer set by the subscription. This function holds significance as it manages the logic that occurs at the conclusion of each month of a subscription.
+
+The advanceMonth function plays a crucial role in the subscription lifecycle, performing several important tasks:
```
Comment (Yaroms) [inline x/subscription/README.md:90]: add the task of paying all providers that served this subscription

Comment (Yaroms) [reply x/subscription/README.md:90]: this is also described in pairing

Comment (shleikes) [reply x/subscription/README.md:90]: Added in 1898ee2ce8245519bcd5d0d2b24d8bba5d5e04f0

**Review by Yaroms — APPROVED** [review]
